### ISSUE TITLE
Change in behavior when (starts-with|ends-with|substr-eq|contains)ing a Junction

### ISSUE THREAD

**Issue by MasterDuke17** [rakudo/rakudo#2719]
```
[12:42] <MasterDuke> m: my @a = ("a".."k").pick(*) xx 100; my $b; for @a -> $a { $b = $a.starts-with("b" | "j") }; say so $b  # what am i doing wrong here?
[12:42] <+camelia> rakudo-moar 616b757a4: OUTPUT: «Type check failed for return value; expected Bool:D but got Junction (any(Bool::False, Bool...)␤  in block <unit> at <tmp> line 1␤␤»
[12:45] <AlexDaniel> oooh…
[12:45] <AlexDaniel> c: 2018.12 my @a = ("a".."k").pick(*) xx 100; my $b; for @a -> $a { $b = $a.starts-with("b" | "j") }
[12:45] <+committable6> AlexDaniel, ¦2018.12: «»
[12:45] <AlexDaniel> c: HEAD my @a = ("a".."k").pick(*) xx 100; my $b; for @a -> $a { $b = $a.starts-with("b" | "j") }
[12:45] <+committable6> AlexDaniel, ¦HEAD(616b757): «Type check failed for return value; expected Bool:D but got Junction (any(Bool::False, Bool...)␤  in block <unit> at /tmp/2fxq5B351T line 1␤␤ «exit code = 1»»
[12:45] <AlexDaniel> MasterDuke: I don't think you're doing anything wrong
[12:45] <AlexDaniel> bisect: old=2018.12 my @a = ("a".."k").pick(*) xx 100; my $b; for @a -> $a { $b = $a.starts-with("b" | "j") }
[12:45] <+timotimo> i'd expect the return value of the loop body has to be an Any
[12:45] <+bisectable6_> AlexDaniel, Bisecting by exit code (old=2018.12 new=616b757). Old exit code: 0
[12:45] <+bisectable6_> AlexDaniel, bisect log: https://gist.github.com/53c4d742899c8d64d64d5f66b223f632
[12:45] <+bisectable6_> AlexDaniel, (2018-12-28) https://github.com/rakudo/rakudo/commit/2ce79561d9ad9a0aea707ada4bd64b7d8fcbebc6
[12:46] <AlexDaniel> yeah
[12:46] <AlexDaniel> that's what I was looking for
[12:46] <+timotimo> oh, huh
[12:46] <+timotimo> the starts-with sub takes a capture, so the autothread the junctions part doesn't trigger there, right
[12:47] <AlexDaniel> m: say ‘foo’.starts-with(‘a’ | ‘b’)
[12:47] <+camelia> rakudo-moar 616b757a4: OUTPUT: «any(False, False)␤»
[12:47] <AlexDaniel> but why does that work?
[12:47] <AlexDaniel> m: say <foo bar>.starts-with(‘a’ | ‘b’)
[12:47] <+camelia> rakudo-moar 616b757a4: OUTPUT: «Type check failed for return value; expected Bool:D but got Junction (any(Bool::False, Bool...)␤  in block <unit> at <tmp> line 1␤␤»
[12:47] <AlexDaniel> ahhh I see, only on Cool
[12:47] <+timotimo> m: my @a = ("a".."k").pick(*) xx 100; my $b; for @a -> $a --> Mu { $b = $a.starts-with("b" | "j") }; say so $b
[12:47] <+camelia> rakudo-moar 616b757a4: OUTPUT: «Type check failed for return value; expected Bool:D but got Junction (any(Bool::False, Bool...)␤  in block <unit> at <tmp> line 1␤␤»
[12:47] <AlexDaniel> MasterDuke: yeah btw you're calling .starts-with on a list
[12:48] <MasterDuke> ah, right. AlexDaniel++
[12:48] <AlexDaniel> I'm pretty sure that needs to be reverted
[12:48] <AlexDaniel> MasterDuke: can you file a ticket?
[12:49] <AlexDaniel> maybe not fully, but for anything that can potentially take a Junction as an argument
[12:54] <AlexDaniel> m: <foo %s bar>.printf(‘abc’ | ‘bca’)
[12:54] <+camelia> rakudo-moar 616b757a4: OUTPUT: «This representation (Null) does not support elems (for type VMNull)␤  in block <unit> at <tmp> line 1␤␤»
[12:54] <AlexDaniel> interesting
[12:54] <AlexDaniel> c: 2018.12 <foo %s bar>.printf(‘abc’ | ‘bca’)
[12:54] <+committable6> AlexDaniel, ¦2018.12: «This representation (Null) does not support elems (for type VMNull)␤  in block <unit> at /tmp/uU1_VVBdCG line 1␤␤ «exit code = 1»»
[12:54] <AlexDaniel> c: all <foo %s bar>.printf('abc' | 'bca')
[12:54] <AlexDaniel> probably never worked
[12:55] <+committable6> AlexDaniel, https://gist.github.com/66f6be2225f8e8cf106cc36b0b45ef8a
[12:55] <AlexDaniel> somehow only these are affected: starts-with ends-with substr-eq contains
```

**Comment by AlexDaniel** [comment]
Obligatory ping: @lizmat

**Comment by lizmat** [comment]
Fixed with 010587db50 , don't feel like spending time to find out which return signatures would work always and which ones don't.  We need to get a release out of the door.

**Comment by lizmat** [comment]
Tests added with https://github.com/Raku/roast/commit/3772d8497b
